adam/curriculum/preposition_curriculum.py

```diff
@@ -0,0 +1,726 @@
+from itertools import chain
+from typing import Iterable
+
+from immutablecollections import immutableset
+from more_itertools import flatten
+
+from adam.axes import HorizontalAxisOfObject, FacingAddresseeAxis
+from adam.curriculum.phase1_curriculum import (
+    Phase1InstanceGroup,
+    standard_object,
+    phase1_instances,
+)
+from adam.ontology import IN_REGION, IS_ADDRESSEE, IS_SPEAKER, THING
+from adam.ontology.phase1_ontology import (
+    BALL,
+    BOOK,
+    BOX,
+    TABLE,
+    on,
+    GAILA_PHASE_1_ONTOLOGY,
+    strictly_above,
+    inside,
+    WATER,
+    JUICE,
+    CUP,
+    MOM,
+    DAD,
+    CAN_HAVE_THINGS_RESTING_ON_THEM,
+    bigger_than,
+    HAS_SPACE_UNDER,
+    IS_BODY_PART,
+    HOLLOW,
+)
+from adam.ontology.phase1_spatial_relations import Region, PROXIMAL, Direction
+from adam.random_utils import RandomChooser
+from adam.relation import Relation
+from adam.situation.templates.phase1_templates import (
+    Phase1SituationTemplate,
+    sampled,
+    TemplateObjectVariable,
+    object_variable,
+)
+
+_CHOOSER = RandomChooser.for_seed(0)
+
+
+def _make_background(
+    salient: Iterable[TemplateObjectVariable],
+    all_objects: Iterable[TemplateObjectVariable],
+) -> Iterable[TemplateObjectVariable]:
+    return immutableset(object_ for object_ in all_objects if object_ not in salient)
+
+
+def _on_templates(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+) -> Phase1SituationTemplate:
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-on-{ground.handle}",
+        salient_object_variables=[figure, ground],
+        background_object_variables=background,
+        asserted_always_relations=[on(figure, ground)],
+    )
+
+
+def _beside_template(
+    figure: TemplateObjectVariable,
+    ground: TemplateObjectVariable,
+    background: Iterable[TemplateObjectVariable],
+    *,
+    is_right: bool,
+) -> Phase1SituationTemplate:
+    direction_str = "right" if is_right else "left"
+    return Phase1SituationTemplate(
+        f"preposition-training-{figure.handle}-beside-{ground.handle}-{direction_str}",
+        salient_object_variables=[figure, ground],
+        background_object_variables=background,
+        asserted_always_relations=[
+            Relation(
+                IN_REGION,
+                figure,
+                Region(
+                    ground,
+                    distance=PROXIMAL,
+                    direction=Direction(
+                        positive=is_right,
+                        relative_to_axis=HorizontalAxisOfObject(ground, index=0),
```

Comment:
Should this be relative to the horizontal axis of the object itself, rather than the ground?

Reply:
This mirrors the way `beside` was constructed in the original curriculum -- That said, I'm open to the change suggested.

Reply:
Make a separate issue for us to think about it later, on the M3 cleanup milestone

Reply:
https://github.com/isi-vista/adam/issues/377